File: work/arc083_d.cpp
#include<iostream>
#include<vector>
using namespace std;

typedef long long ll;

int main(){
  ll n;
  cin>>n;
  vector<vector<ll>> d = vector<vector<ll>>(n, vector<ll>(n,0));
  for(ll i=0;i<n;++i){
    for(ll j=0;j<n;++j){
      cin>>d[i][j];
    }
  }
  for(ll k=0;k<n;++k){
    for(ll i=0;i<n;++i){
      for(ll j=0;j<n;++j){
        if (d[i][j]>d[i][k]+d[k][j]){
          cout<<-1<<endl;
          return 0;
        }
      }
    }
  }
  ll res = 0;
  for(ll i=0;i<n;++i){
    for(ll j=i+1;j<n;++j){
      bool flag = true;
      for(ll k=0;k<n;++k){
        if(i==k || j==k){
          continue;
        }
        if(d[i][j]==d[i][k]+d[k][j]){
          flag = false;
          break;
        }
      }
      if(flag){
        res += d[i][j];
      }
    }
  }
  cout<<res<<endl;
}

File: work/arc083_d.py
# -*- coding: utf-8 -*-
n = int(input())
d = []
for i in range(n):
    line = [int(x) for x in input().split()]
    d.append(line)

for k in range(n):
    for i in range(n):
        for j in range(n):
            if d[i][j]>d[i][k]+d[k][j]:
                print(-1)
                exit()

res = 0
for i in range(i):
    for j in range(i+1,n):
        flag = True
        for k in range(n):
            if i==k or j==k:
                continue
            if d[i][j]==d[i][k]+d[k][j]:
                flag = False
                break
        if flag:
            res += d[i][j]
print(res)
